File: 557-reverse-words-in-a-string-iii/rust/src/main.rs
struct Solution;
impl Solution {
    pub fn reverse_words(s: String) -> String {
        let mut res = String::new();
        res.reserve(s.len());
        let mut word = String::new();
        for c in s.chars() {
            if c == ' ' {
                word = word.chars().rev().collect();
                res += &word;
                res.push(' ');
                word.clear();
            } else {
                word.push(c);
            }
        }
        word = word.chars().rev().collect();
        res += &word;
        res
    }
}

struct Solution2;
impl Solution2 {
    pub fn reverse_words(s: String) -> String {
        let mut char_vec: Vec<char> = s.chars().collect();
        let n = char_vec.len();
        let mut i = 0;
        while i < n {
            let mut j = i;
            while j < n && char_vec[j] != ' ' {
                j += 1;
            }
            let next = j;
            j -= 1;
            while i < j {
                char_vec.swap(i, j);
                i += 1;
                j -= 1;
            }
            i = next + 1;
        }
        char_vec.iter().collect()
    }
}

struct Solution3;
impl Solution3 {
    pub fn reverse_words(s: String) -> String {
        s.split(" ")
            .map(|x| x.chars().rev().collect::<String>())
            .collect::<Vec<_>>()
            .join(" ")
    }
}

struct Solution4;
impl Solution4 {
    pub fn reverse_words(s: String) -> String {
        s.chars()
            .rev()
            .collect::<String>()
            .split_whitespace()
            .rev()
            .collect::<Vec<_>>()
            .join(" ")
    }
}

fn main() {
    {
        let res = Solution::reverse_words("Let's take LeetCode contest".to_string());
        print!("{}\n", res);
    }
    {
        let res = Solution2::reverse_words("Let's take LeetCode contest".to_string());
        print!("{}\n", res);
    }
    {
        let res = Solution3::reverse_words("Let's take LeetCode contest".to_string());
        print!("{}\n", res);
    }
    {
        let res = Solution4::reverse_words("Let's take LeetCode contest".to_string());
        print!("{}\n", res);
    }
}

File: 557-reverse-words-in-a-string-iii/cpp/main.cpp
#include <algorithm>
#include <iostream>

using namespace std;

class Solution {
public:
  string reverseWords(string s) {
    string res;
    res.reserve(s.size());
    string word;
    for (auto c : s) {
      if (c == ' ') {
        reverse(word.begin(), word.end());
        res += word + c;
        word.clear();
      } else {
        word += c;
      }
    }
    reverse(word.begin(), word.end());
    res += word;

    return res;
  }
};

// OJ: https://leetcode.com/problems/reverse-words-in-a-string-iii
// Author: github.com/lzl124631x
// Time: O(N)
// Space: O(1)
class Solution2 {
public:
  string reverseWords(string s) {
    int N = s.size();
    for (int i = 0; i < N; ++i) {
      int j = i;
      while (j < N && s[j] != ' ') ++j;
      int next = j;
      j-=1;

      for (--j; i < j;) swap(s[i++], s[j--]);
      i = next;
    }
    return s;
  }
};

// OJ: https://leetcode.com/problems/reverse-words-in-a-string-iii
// Author: github.com/lzl124631x
// Time: O(N)
// Space: O(1)
class Solution3 {
public:
  string reverseWords(string s) {
    for (int i = 0, begin = 0, N = s.size(); i <= N; ++i) {
      if (i == N || s[i] == ' ') {
        reverse(s.begin() + begin, s.begin() + i);
        begin = i + 1;
      }
    }
    return s;
  }
};

int main()
{
  try
  {
    {
      Solution s;
      const auto res = s.reverseWords("Let's take LeetCode contest");
      cout << res << endl;
      int j = 0;
    }
    {
      Solution s;
      const auto res = s.reverseWords("God Ding");
      cout << res << endl;
      int j = 0;
    }
  }
  catch ( const exception& e )
  {
    cerr << e.what();
  }

  return 0;
}